### ISSUE TITLE
opaqueGenericParameters breaks compilation in some instances

### ISSUE THREAD

**Issue by noremac** [nicklockwood/SwiftFormat#1269]
I just started trying out the `0.50.0` release and have noticed two issues with `opaqueGenericParameters`.

### Before

```swift
struct Foo<Value> {
    func bar<V, R>(
        _ value: V,
        _ work: () -> R
    ) -> R
    where Value == @Sendable () -> V,
          V: Sendable
    {
        work()
    }
}
```

### After

```swift
struct Foo<Value> {
    func bar<R>(
        _ value: some Sendable,
        _ work: () -> R
    ) -> R
        where Value == @Sendable () -> V,
        {
        work()
    }
}
```

`opaqueGenericParameters` is not factoring in the constraint of `where Value == @Sendable () -> V` and is deleting the `V` generic parameter.

### Crash

Additionally, this (similar) code will cause swift format to crash:

```swift
struct Foo<Value> {
    func bar<V>(_ value: V) where Value == @Sendable () -> V, V: Sendable {}
}
```

**Comment by nicklockwood** [comment]
@noremac thanks, I'll get this fixed asap

**Comment by Lutzifer** [comment]
Another instance of opaqueGenericParameters breaking code:

```diff
  func loggedKeypath<T: CustomStringConvertible>(
-    by paths: KeyPath<Element, T>...,
-  func loggedKeypath(
+    by paths: KeyPath<Element, some CustomStringConvertible>...,
+   actionKeyword: UserActionKeyword,
```

Error:

```
'some' types are only permitted in properties, subscripts, and functions
```

**Comment by nicklockwood** [comment]
@calda I could use your help with this. I've fixed the crash (which was a range issue to do with `->` appearing after the `where` keyword, however the underlying issue for the other breakages is that the logic fails to consider the possibility of the generic type  appearing on the right hand side of a where condition - it only looks for them on the left.

So for example, this case is completely broken because the rule doesn't see the `V` after the `==`:

```swift
func foo<U, V>(_ u: U, _ v: V) where U == V {
    print(u, v)
}
```

**Comment by nicklockwood** [comment]
@Lutzifer this bug seems unrelated to the first - would you mind opening a separate issue for it?

**Comment by Lutzifer** [comment]
@nicklockwood Extracted to https://github.com/nicklockwood/SwiftFormat/issues/1275

**Comment by calda** [comment]
Thanks for fixing the crash! I'll fix the logic error.

**Comment by calda** [comment]
Fixed in #1277, thanks for the report!
